lib/rules-generator.js

```diff
@@ -361,37 +407,49 @@ util.methods(Generator, {
       if (typeof exp.accessor == 'string') {
         accessor = exp.accessor;
       } else {
-        accessor = subExpression(exp.accessor);
+        accessor = valueExpression(subExpression(exp.accessor));
       }
-      accessor = valueExpression(accessor);
-      // snapshot references use child() wrapper - EXCEPT for built-in methods
-      // TODO: Resolve ambiguity between built-ins and user-defined property names
-      // (use functions instead)?
-      if (base.valueType == 'Snapshot' && snapshotMethods.indexOf(accessor) == -1) {
-        exp = ast.snapshotChild(base, accessor);
-      } else {
-        if (replaceExp()) {
-          exp.base = base;
-          exp.accessor = accessor;
+      // snapshot references use child() wrapper - EXCEPT for reserved methods
+      // TODO: Only do this if we are dereferencing a function call OR if
+      // there is no shadowed property in the object.
+      if (base.valueType == 'Snapshot') {
+        if (accessor == 'parent') {
+          return ast.snapshotParent(base);
+        } else if (!util.arrayIncludes(reservedMethods, accessor)) {
+          return ast.snapshotChild(base, accessor);
         }
       }
+      if (util.arrayIncludes(valueMethods, accessor)) {
+        base = valueExpression(base);
+      }
+      if (replaceExp()) {
+        exp.base = base;
+        exp.accessor = accessor;
+      }
       break;
 
     case 'call':
       var ref = subExpression(exp.ref);
-      var fn = this.lookupFunction(ref);
-      if (fn) {
+      var callee = this.lookupFunction(ref);
+      if (callee) {
+        var fn = callee.fn;
+
         if (functionCalls[ref.name]) {
           throw new Error(errors.recursive + " (" + ref.name + ")");
         }
-        if (fn.params.length != exp.args.length) {
+        var callArgs = exp.args;
+        if (callee.self) {
+          callArgs = util.copyArray(callArgs);
+          callArgs.unshift(ast.ensureValue(callee.self));
+        }
+        if (fn.params.length != callArgs.length) {
           this.fatal(errors.mismatchParams + " ( " +
                      ref.name + " expects " + fn.params.length +
-                     "but actually passed " + exp.args.length + ")");
+                     " but actually passed " + callArgs.length + ")");
           break;
```

Comment:
break in the middle of a block is not good syle. Put the happy case first: if (fn.params.length == callArgs.length), and the recursion inside the happy case block, and just have the else unhappy case and reuse the break at the end